File: unreleased/Store/WaveformAnalyzer.swift
import AVFoundation
import Foundation

/// Extracts a normalized amplitude envelope from an audio file.
///
/// Uses `AVAudioFile`, which decodes any supported container (m4a, wav, aiff,
/// mp3, …) into a canonical non-interleaved Float32 buffer. This avoids the
/// `AVAssetReader` sample-rate/channel conversions that silently yield no
/// samples for already-LPCM sources such as WAV.
enum WaveformAnalyzer {

    /// Analyze the file at `url` and return `targetBars` normalized amplitude values (0…1).
    /// Runs off the main actor so it never blocks the UI.
    static func analyze(url: URL, targetBars: Int = 200) async -> [Float] {
        await Task.detached(priority: .userInitiated) {
            (try? Self.extractBars(url: url, targetBars: targetBars)) ?? []
        }.value
    }

    // MARK: - Core extraction (runs on background thread)

    private static func extractBars(url: URL, targetBars: Int) throws -> [Float] {
        let file = try AVAudioFile(forReading: url)
        // processingFormat is always deinterleaved Float32 at the file's native
        // sample rate / channel count — samples are already in -1…1.
        let format = file.processingFormat
        let totalFrames = Int(file.length)
        let channelCount = Int(format.channelCount)
        guard totalFrames > 0, channelCount > 0 else { return [] }

        let framesPerBar = max(1, totalFrames / targetBars)

        var barSumSq = [Double](repeating: 0, count: targetBars)
        var barCount = [Int](repeating: 0, count: targetBars)
        var frameCursor = 0

        let bufferCapacity: AVAudioFrameCount = 65_536
        guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: bufferCapacity) else {
            return []
        }

        while file.framePosition < file.length {
            try file.read(into: buffer)
            let frames = Int(buffer.frameLength)
            if frames == 0 { break }
            guard let channels = buffer.floatChannelData else { break }

            for f in 0..<frames {
                var sumSq = 0.0
                for c in 0..<channelCount {
                    let v = Double(channels[c][f])
                    sumSq += v * v
                }
                let power = sumSq / Double(channelCount)
                let barIdx = min(targetBars - 1, frameCursor / framesPerBar)
                barSumSq[barIdx] += power
                barCount[barIdx] += 1
                frameCursor += 1
            }
        }

        // RMS per bar (float samples are already normalized to 0…1).
        var bars = zip(barSumSq, barCount).map { sumSq, n -> Float in
            guard n > 0 else { return 0 }
            return Float(sqrt(sumSq / Double(n)))
        }

        let peak = bars.max() ?? 0
        if peak > 0 { bars = bars.map { min(1, $0 / peak) } }

        // Clamp very quiet bars to a minimum visible height so silence isn't invisible.
        return bars.map { max(0.05, $0) }
    }
}

File: web/src/lib/waveform-analyzer.ts
/**
 * 1:1 port of the iOS `WaveformAnalyzer`.
 *
 * The iOS pipeline (AVAudioFile → Float32 PCM → per-bar RMS):
 *   1. Decode the file to non-interleaved Float32 PCM.
 *   2. `framesPerBar = max(1, totalFrames / targetBars)`; remainder frames
 *      accumulate into the last bar (`barIdx = min(targetBars-1, f / framesPerBar)`).
 *   3. Per frame: `power = mean(sample² across channels)`.
 *   4. Per bar: `RMS = sqrt(mean(power))` (Double accumulation, like iOS).
 *   5. Normalize by the peak bar.
 *   6. Clamp quiet bars to 0.05 so silence stays visible.
 *
 * Every arithmetic step matches, so lossless sources (WAV/AIFF/FLAC) produce
 * byte-identical waveform payloads. Lossy sources (M4A/MP3) may differ by at
 * most a quantization step depending on the browser's decoder — visually
 * indistinguishable from an iOS import.
 */
export interface AnalyzedAudio {
  duration: number
  /** Normalized amplitude bars (0…1); empty when the browser can't decode the file. */
  waveform: number[]
}

export async function analyzeAudioFile(file: File, targetBars = 200): Promise<AnalyzedAudio> {
  let buffer: AudioBuffer
  try {
    // A fixed 44.1 kHz offline context keeps decoding deterministic across
    // machines (decodeAudioData resamples to the context rate). The envelope
    // math below is frame-count-relative, matching the iOS analyzer.
    const context = new OfflineAudioContext(1, 1, 44_100)
    buffer = await context.decodeAudioData(await file.arrayBuffer())
  } catch {
    // Mirrors iOS: analysis failures still import the track (`waveform ?? []`).
    return { duration: 0, waveform: [] }
  }

  return { duration: buffer.duration, waveform: extractBars(buffer, targetBars) }
}

function extractBars(buffer: AudioBuffer, targetBars: number): number[] {
  const totalFrames = buffer.length
  const channelCount = buffer.numberOfChannels
  if (totalFrames <= 0 || channelCount <= 0) return []

  const channels: Float32Array[] = []
  for (let c = 0; c < channelCount; c++) channels.push(buffer.getChannelData(c))

  const framesPerBar = Math.max(1, Math.floor(totalFrames / targetBars))
  const barSumSq = new Float64Array(targetBars)
  const barCount = new Int32Array(targetBars)

  for (let f = 0; f < totalFrames; f++) {
    let sumSq = 0
    for (let c = 0; c < channelCount; c++) {
      const v = channels[c][f]
      sumSq += v * v
    }
    const power = sumSq / channelCount
    const barIdx = Math.min(targetBars - 1, Math.floor(f / framesPerBar))
    barSumSq[barIdx] += power
    barCount[barIdx] += 1
  }

  // RMS per bar — Math.fround mirrors the iOS casts to 32-bit Float
  // (`Float(sqrt(sumSq / Double(n)))` and Float division during normalization).
  let bars = new Array<number>(targetBars)
  for (let i = 0; i < targetBars; i++) {
    bars[i] = barCount[i] > 0 ? Math.fround(Math.sqrt(barSumSq[i] / barCount[i])) : 0
  }

  const peak = Math.max(...bars)
  if (peak > 0) bars = bars.map((b) => Math.min(1, Math.fround(b / peak)))

  // Clamp very quiet bars to a minimum visible height so silence isn't invisible.
  const minHeight = Math.fround(0.05)
  return bars.map((b) => Math.max(minHeight, b))
}
